ipset.go

```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
 		if debug {
-			debugf("ret=%v", ret)
+			debugf("min=%v", min)
 		}
 	}
 
-	// TODO: possibly update s.in and s.out, if #110 supports that.
+	s.in = min
+	s.out = nil
+}
+
+// Add adds ip to the set s.
+func (s *IPSetBuilder) Add(ip IP) { s.AddRange(IPRange{ip, ip}) }
+
+// AddPrefix adds p's range to s.
+func (s *IPSetBuilder) AddPrefix(p IPPrefix) { s.AddRange(p.Range()) }
 
-	return ret
+// AddRange adds r to s.
+func (s *IPSetBuilder) AddRange(r IPRange) {
+	if !r.Valid() {
+		return
+	}
+	// If there are any removals (s.out), then we need to compact the set
+	// first to get the order right.
+	if len(s.out) > 0 {
+		s.normalize()
+	}
+	s.in = append(s.in, r)
+}
+
+// AddSet adds all ranges in b to s.
+func (s *IPSetBuilder) AddSet(b *IPSet) {
+	for _, r := range b.rr {
+		s.AddRange(r)
+	}
+}
+
+// Remove removes ip from the set s.
+func (s *IPSetBuilder) Remove(ip IP) { s.RemoveRange(IPRange{ip, ip}) }
+
+// RemoveFreePrefix removes and returns a Prefix of length bits from the IPSetBuilder.
+func (s *IPSetBuilder) RemoveFreePrefix(bitLen uint8) (p IPPrefix, ok bool) {
+	s.normalize()
+
+	var bestFit IPPrefix
+	for _, r := range s.in {
+		for _, prefix := range r.Prefixes() {
+			if prefix.Bits > bitLen {
+				continue
+			}
+			if bestFit.IP.IsZero() || prefix.Bits > bestFit.Bits {
+				bestFit = prefix
+				if bestFit.Bits == bitLen {
+					// exact match, done.
+					break
+				}
+			}
+		}
+	}
+
+	if bestFit.IP.IsZero() {
+		return IPPrefix{}, false
+	}
+
+	prefix := IPPrefix{IP: bestFit.IP, Bits: bitLen}
+	s.RemovePrefix(prefix)
+	return prefix, true
+}
+
+// RemovePrefix removes p's range from s.
+func (s *IPSetBuilder) RemovePrefix(p IPPrefix) { s.RemoveRange(p.Range()) }
+
+// RemoveRange removes r from s.
+func (s *IPSetBuilder) RemoveRange(r IPRange) {
+	if r.Valid() {
+		s.out = append(s.out, r)
+	}
+}
+
+// RemoveSet removes all ranges in o from s.
+func (s *IPSetBuilder) RemoveSet(o *IPSet) {
+	for _, r := range o.rr {
+		s.RemoveRange(r)
+	}
+}
+
+// removeBuilder removes all IPs in b from s.
+func (s *IPSetBuilder) removeBuilder(b *IPSetBuilder) {
+	b.normalize()
+	for _, r := range b.in {
+		s.RemoveRange(r)
+	}
+}
+
+// Complement updates s to contain the complement of its current
+// contents.
+func (s *IPSetBuilder) Complement() {
+	s.normalize()
+	s.out = s.in
+	s.in = []IPRange{
+		IPPrefix{IP: IPv4(0, 0, 0, 0), Bits: 0}.Range(),
+		IPPrefix{IP: IPv6Unspecified(), Bits: 0}.Range(),
+	}
+}
+
+// Intersect updates s to the set intersection of s and b.
+func (s *IPSetBuilder) Intersect(b *IPSet) {
+	var o IPSetBuilder
+	o.Complement()
+	o.RemoveSet(b)
+	s.removeBuilder(&o)
+}
+
+func discardf(format string, args ...interface{}) {}
+
+// debugf is reassigned by tests.
+var debugf = discardf
+
+func (s *IPSetBuilder) IPSet() *IPSet {
+	s.normalize()
+	return &IPSet{
+		rr: append([]IPRange{}, s.in...),
+	}
+}
+
+// IPSet represents a set of IP addresses.
+//
+// The zero value is a valid value representing a set of no IPs.
```

Comment:
Add that it's safe for concurrent use.

Reply:
Add

// Use IPSetBuilder to construct IPSets.